In [1]:
!pip install otter-grader

Defaulting to user installation because normal site-packages is not writeable
Looking in indexes: https://packagemanager.posit.co/pypi/latest/simple


In [2]:
# Initialize Otter
import otter
grader = otter.Notebook("hwk1-task2-corals.ipynb")

ModuleNotFoundError: No module named 'otter'

# Task 2: Exploring coral diversity data

## Instructions 

- First, update the following cell to have a link to *your* Homework 1 GitHub repository:


**UPDATE THIS LINK:**

[https://github.com/cloover7/eds220-hwk1.git](https://github.com/cloover7/eds220-hwk1.git)

- This task covers topics on the lesson on [subsetting `pandas.DataFrames`](https://meds-eds-220.github.io/MEDS-eds-220-course/book/chapters/lesson-3-pandas-subsetting/lesson-3-pandas-subsetting.html). The best strategy to solve it is to get together with your classmates and work on it together after that lecture.

- Review the [complete rubric for this task](https://docs.google.com/document/d/131OnUDOLpOl94SlaMU9nGqT998ABrjFS/edit?usp=sharing&ouid=111507336322611936333&rtpof=true&sd=true) before starting.

- **Make at least 5 commits at moments you think are adequate.** We'll check your repository and view the commit history.

- **Add comments for all your code.** Err on the side of commenting too much for now. Comments should follow best practices.



<!-- BEGIN QUESTION -->


## 1
For this task we are going to use data about  [Western Indian Ocean Coral Diversity (McClanahan, 2023)](https://knb.ecoinformatics.org/view/doi%3A10.5063%2FF1K35S3H) from the Knowledge Network for Biocomplexity (KNB) data repository. 

a. Navigate to the dataset's link and explore the data package. 
 
b. Use the next markdown cell to add a brief description of the dataset, including a citation, date of access, and a link to the archive. 


The Western Indian Ocean Coral Diversity dataset was compiled to make machine learning models that predict coral health in the Western Indian Ocean. Data was collected between 1999 and 2022, and the goal is to predict up to 2050. Each row represents one observation of a 6km reef cell at a specific place and time. Columns contain various variables associated with each observation, such as temperature, depth, dissolved oxygen, and salinity.

Date of access: Oct 4, 2026

[Data Source](https://knb.ecoinformatics.org/view/doi%3A10.5063%2FF1K35S3H#urn%3Auuid%3Ab9d486bb-ea8d-442d-8287-c66c0ce18425)

<!-- END QUESTION -->

<!-- BEGIN QUESTION -->

## 2
In the next code cell, import the `pandas` package using standard abbreviation in a code cell.

Read in the data from the URL using the `pd.read_csv` function and store it as `coral_div` like this:

```
# read in data
coral_div = pd.read_csv('the URL goes here')
```

In [3]:
import pandas as pd
coral_div = pd.read_csv('data/coral_data.csv')
coral_div.head()

,PU_ID,PU_Lat,PU_Lon,Number_of_genera,Country,Ecoregion,Management,Habitat,Observer,Depth,...,SST_bimodality,SST_Skewness,ClimateStressModel,Netflow,Indegree,Outdegree,Retention,min_yr,max_yr,no_replicates
0,7,39.076711,-5.490011,35,Tanzania,East African Coral Coast,UnfishedLow,Crest,Tim,11.0,...,0.462556,-0.128136,0.596375,0.144498,98.25,74.75,14.00,2019,2019,1
1,7,39.076711,-5.490011,22,Tanzania,East African Coral Coast,UnfishedLow,Lagoon/Back reef,Nyawira,2.0,...,0.462556,-0.128136,0.596375,0.144498,98.25,74.75,14.00,2010,2015,2
2,7,39.076711,-5.490011,25,Tanzania,East African Coral Coast,UnfishedLow,Slope,Nyawira,4.0,...,0.462556,-0.128136,0.596375,0.144498,98.25,74.75,14.00,2015,2015,1
3,7,39.076711,-5.490011,26,Tanzania,East African Coral Coast,UnfishedLow,Slope,Nyawira,3.0,...,0.462556,-0.128136,0.596375,0.144498,98.25,74.75,14.00,2016,2019,2
4,10,39.054253,-5.512514,27,Tanzania,East African Coral Coast,UnfishedLow,Lagoon/Back reef,Nyawira,1.5,...,0.461297,-0.134548,0.629000,0.011721,112.75,51.50,183.75,2010,2019,3


<!-- END QUESTION -->

## 3

Use attributes and methods of the `coral_div` data frame to obtain the following information about it:

Use the code cells marked a-e. For each, add a comment explaining what is the ouput.

a. how many rows and columns does it have? Store your answers in the `rows` and `columns` variables. 

In [4]:
#a
rows = coral_div.shape[0]
columns = coral_div.shape[1]
rows

575

In [5]:
grader.check("q3_a")

NameError: name 'grader' is not defined

b. What are the column names? Store your answer in the `coral_div_columns` variable.

In [6]:
#b
coral_div_columns = coral_div.columns
coral_div_columns

Index(['PU_ID', 'PU_Lat', 'PU_Lon', 'Number_of_genera', 'Country', 'Ecoregion',
       'Management', 'Habitat', 'Observer', 'Depth', 'Coralcover',
       'SSTRateofRise', 'PARmax', 'Calcite', 'Dis_oxygen', 'mean.waveNRJ',
       'Salinity_mean', 'andrello_reef_value', 'andrello_nutrients',
       'mean.npp', 'TT_market_hrs', 'TT_pop_hrs', 'cumDHW', 'SST_Kurtosis',
       'SST_Median', 'ChlorA_median', 'andrello_sediments', 'Current_vel_mean',
       'Diff_attenuation', 'Grav_NP', 'Grav_NC', 'PH', 'SST_bimodality',
       'SST_Skewness', 'ClimateStressModel', 'Netflow', 'Indegree',
       'Outdegree', 'Retention', 'min_yr', 'max_yr', 'no_replicates'],
      dtype='object')

In [ ]:
grader.check("q3_b")

c. what are the data types of the columns? Store your answer in the `coral_div_types`.

In [7]:
#c
coral_div_types = coral_div.dtypes
coral_div_types

PU_ID                    int64
PU_Lat                 float64
PU_Lon                 float64
Number_of_genera         int64
Country                 object
Ecoregion               object
Management              object
Habitat                 object
Observer                object
Depth                  float64
Coralcover               int64
SSTRateofRise          float64
PARmax                 float64
Calcite                float64
Dis_oxygen             float64
mean.waveNRJ           float64
Salinity_mean          float64
andrello_reef_value      int64
andrello_nutrients     float64
mean.npp               float64
TT_market_hrs          float64
TT_pop_hrs             float64
cumDHW                 float64
SST_Kurtosis           float64
SST_Median             float64
ChlorA_median          float64
andrello_sediments     float64
Current_vel_mean       float64
Diff_attenuation       float64
Grav_NP                float64
Grav_NC                float64
PH                     float64
SST_bimo

In [ ]:
grader.check("q3_c")

d. display the first 8 rows with all the column names. For this one, you will need to udpate a `pandas` option (HINT: check the first discussion section). Store your answer in the `first_8_rows` variable.

In [8]:
#d
first_8_rows = coral_div.head(8)

In [9]:
# View the dataframe 
first_8_rows

,PU_ID,PU_Lat,PU_Lon,Number_of_genera,Country,Ecoregion,Management,Habitat,Observer,Depth,...,SST_bimodality,SST_Skewness,ClimateStressModel,Netflow,Indegree,Outdegree,Retention,min_yr,max_yr,no_replicates
0,7,39.076711,-5.490011,35,Tanzania,East African Coral Coast,UnfishedLow,Crest,Tim,11.0,...,0.462556,-0.128136,0.596375,0.144498,98.25,74.75,14.00,2019,2019,1
1,7,39.076711,-5.490011,22,Tanzania,East African Coral Coast,UnfishedLow,Lagoon/Back reef,Nyawira,2.0,...,0.462556,-0.128136,0.596375,0.144498,98.25,74.75,14.00,2010,2015,2
2,7,39.076711,-5.490011,25,Tanzania,East African Coral Coast,UnfishedLow,Slope,Nyawira,4.0,...,0.462556,-0.128136,0.596375,0.144498,98.25,74.75,14.00,2015,2015,1
3,7,39.076711,-5.490011,26,Tanzania,East African Coral Coast,UnfishedLow,Slope,Nyawira,3.0,...,0.462556,-0.128136,0.596375,0.144498,98.25,74.75,14.00,2016,2019,2
4,10,39.054253,-5.512514,27,Tanzania,East African Coral Coast,UnfishedLow,Lagoon/Back reef,Nyawira,1.5,...,0.461297,-0.134548,0.629000,0.011721,112.75,51.50,183.75,2010,2019,3
5,10,39.054253,-5.512514,29,Tanzania,East African Coral Coast,UnfishedLow,Lagoon/Back reef,Tim,1.5,...,0.461297,-0.134548,0.629000,0.011721,112.75,51.50,183.75,2015,2015,1
6,15,39.054253,-5.535017,19,Tanzania,East African Coral Coast,Restricted,Crest,Tim,8.0,...,0.461297,-0.134548,0.595600,0.011721,112.75,51.50,183.75,2019,2019,1
7,39,39.009337,-5.647518,28,Tanzania,East African Coral Coast,UnfishedLow,Lagoon/Back reef,Nyawira,3.0,...,0.464874,-0.158601,0.626000,-0.063087,96.50,40.75,35.75,2015,2019,3


In [ ]:
grader.check("q3_d")

<!-- BEGIN QUESTION -->

e. Write code to find out some other inital information you are interested in. 

In [ ]:
# e
coral_div['Habitat'].unique()

<!-- END QUESTION -->

<!-- BEGIN QUESTION -->

## 4

Some of the columns have data type `int64` and some have data type `float64`. What is the difference between these `int64` and `float64`?


`int64` stands for integer so these are exact numbers with no decimals or fractions such as 3. `float64` is for floats, which may include decimals such as 1.25.

<!-- END QUESTION -->

<!-- BEGIN QUESTION -->

## 5

The `Country` column indicates the country where the coral reef is located. Explain, in your own words, what is the ouput of `coral_div['Country']` and `coral_div['Country'].unique()`. 

`coral_div['Country']` will list all 575 values under the Country column. In other words, it outputs the entire column.  `coral_div['Country'].unique()` will only output each unique value in the column once.

<!-- END QUESTION -->

## 6

Check the dataset's [metadata file README_md.doc](https://knb.ecoinformatics.org/view/doi:10.5063/F1K35S3H) to find which column has information about who made an observation. How many observers collected this data? HINT: Check this [method for `pandas.DataFrame`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.nunique.html). Store your answer in the `unique_observers` variable.

In [10]:
unique_observers =  coral_div['Observer'].nunique()

In [ ]:
grader.check("q6")

## 7

The `Coralcover` column indicates the percent of hard coral cover in each 6 km reef cell surveyed. 

<!-- BEGIN QUESTION -->

a)  Describe, in detail,  what kind of object `coral_div['Coralcover']>= 50` is.

In [13]:
coral_div['Coralcover']>= 50

0      False
1      False
2      False
3      False
4      False
       ...  
570    False
571    False
572    False
573    False
574    False
Name: Coralcover, Length: 575, dtype: bool

In [11]:
type(coral_div['Coralcover'] >= 50)

pandas.core.series.Series

In [12]:
type(coral_div)

pandas.core.frame.DataFrame

This would be a pandas series, which is what pandas data frames are made up of. This is because `coral_div['Coralcover']>= 50` outputs a 

<!-- END QUESTION -->

<!-- BEGIN QUESTION -->

b) What data is `coral_div[coral_div['Coralcover'] >= 50]` selecting?

 This includes all rows and columns for which the value of the 'Coralcover' column is greater than or equal to 50.

<!-- END QUESTION -->

## 8

Use code to find if there are there any sites with a 30% coral cover in Tanzania. Store your answer in the `coral_cover` variable.

In [ ]:
coral_cover = coral_div[(coral_div['Country'] == 'Tanzania') & (coral_div['Coralcover'] == 30)]

In [ ]:
# View the dataframe
coral_cover

In [ ]:
grader.check("q8")

## 9

We are interested in looking at the PH of sea water (`PH` column),  mean salinity (`Salinity_mean` column), and depth of survey location in meters (`Depth` column). Subset the data frame by selecting only these column. Store your answer in the `filtered_data` variable

In [ ]:
filtered_data = coral_div.loc[:, ['PH', 'Salinity_mean', 'Depth']]



In [ ]:
# View the dataframe
filtered_data.head()

In [ ]:
grader.check("q9")

## 10
The `max_yr` column indicates the year of last sampling at a given site. Select the PH, salinity, and depth for sites where sampling ended in 2017. Store your answer in the `sampling_2017` variable. `Sampling_2017` should include the following columns in the following order: ` PH, Salinity_mean, Depth, max_yr`. 


In [ ]:
sampling_2017 = coral_div[coral_div['max_yr'] == 2017].loc[:, ['PH', 'Salinity_mean', 'Depth']]

In [ ]:
# View the dataframe 
sampling_2017

In [ ]:
grader.check("q10")

## 11
Use `iloc` to select rows 95 through 100 (including 100) in the 17th, 20th, and 28th columns and assign these values to a new variable called `subset`. View the new `subset` dataframe.

HINT: the ouput columns should be `Salinity_mean`, `mean.npp` and `Current_vel_mean` (in that order!).

In [ ]:
subset = coral_div.iloc[95:101, [16, 19, 27]]

In [ ]:
# View the dataframe
subset

In [ ]:
grader.check("q11")

<!-- BEGIN QUESTION -->

## 12 

Run `subset.isna()`. Explain what this code does and what the output represents.

In [ ]:
subset.isna()

This code checks if each cell in the data frame is a null value such as 'NaN'. It outputs a data frame of the same dimensions filled with Trues and Falses based on whether that cell is a NaN or not.

<!-- END QUESTION -->

<!-- BEGIN QUESTION -->

## 13
Run `subset.isna().any()`. Explain in your own words what this code does and what the output represents. HINT: check the [documentation of `pandas.Sries.any`](https://pandas.pydata.org/docs/reference/api/pandas.Series.any.html).


In [ ]:
subset.isna().any()

This code takes in a dataframe and evaluates whether there are any null values in each column. It outputs a series containing each column name and a True or False assigned to it based on whether there are any NaNs in that column.

<!-- END QUESTION -->

## 14 

Select the rows in `subset` that have NA in the `Current_vel_mean` column. Store your answer in the `subset_na` variable.

In [ ]:
subset_na = subset[subset['Current_vel_mean'].isna()]

In [ ]:
# View the dataframe 
subset_na

In [ ]:
grader.check("q14")

---

To double-check your work, the cell below will rerun all of the autograder tests.

In [ ]:
grader.check_all()